model-api/xml-model/src/main/java/org/camunda/bpm/model/xml/impl/validation/ModelValidationResultsImpl.java

```diff
@@ -75,9 +74,50 @@ public void write(StringWriter writer, ValidationResultFormatter formatter) {
     }
   }
 
+  @Override
+  public void write(StringWriter writer, ValidationResultFormatter formatter, int maxSize) {
+    int printedCount = 0;
+    int previousLength = 0;
+    for (var entry : collectedResults.entrySet()) {
+      var element = entry.getKey();
+      var results = entry.getValue();
+
+      formatter.formatElement(writer, element);
+
+      for (var result : results) {
+        formatter.formatResult(writer, result);
+
+        // Size and Length are not necessarily the same, depending on the encoding of the string.
+        int currentSize = writer.getBuffer().toString().getBytes().length;
+        int currentLength = writer.getBuffer().length();
+        if (!canAccommodateResult(maxSize, currentSize, printedCount)) {
+          writer.getBuffer().setLength(previousLength);
+          writer.append(suffixWithNumberOfRemainingResults(printedCount));
+          return;
+        }
+        printedCount++;
+        previousLength = currentLength;
+      }
+    }
+  }
+
+  private boolean canAccommodateResult(int maxSize, int currentSize, int printedCount) {
+    boolean lastItemToPrint = printedCount == errorCount + warningCount - 1;
+    if (lastItemToPrint && currentSize <= maxSize) {
+      return true;
+    }
+    
+    int suffixLength = suffixWithNumberOfRemainingResults(printedCount).length();
+    return currentSize + suffixLength <= maxSize;
+  }
+
+  private String suffixWithNumberOfRemainingResults(int printedCount) {
+    int remaining = errorCount + warningCount - printedCount;
+    return String.format(" and %d more errors and/or warnings", remaining);
+  }
```

Comment:
🔧 I believe this formatting should be implemented by the `ValidationErrorFormatter`. For example, I was wondering how the space ` ` at the start of this message would play together with the other formatted text, but I can only find that out by navigating to the formatter. I think we can improve this by moving the method to the formatter.

Reply:
I agree with you in concept, and I thought about it too 🙏 

I could move it indeed to the formatter to and declare a method `void formatSuffixWithOmittedResults(StringWriter writer, int omitted)`.

But then I'd miss a way to know the length of this suffix in here during writing, and I'd not be able to know if I accommodate space for it in the maxsize.

We could then choose to have the max size to cap the printable errors only, and consider the suffix as an accepted extra bytes. But then we would need to find a way to make this clear in documentation

Reply:
Ah I see, sounds like you need to split formatting from determining how long something will be. Perhaps you can add a method that returns the length of the suffix when it would be added now? So you'll need two new methods:
- one to determine the length of the suffix before adding it
- one to add the formatted suffix

Reply:
Good idea! I will try to refactor it this way and see!
Thanks Nico!